clang/include/clang/AST/Type.h

```diff
@@ -4639,6 +4644,312 @@ class FunctionNoProtoType : public FunctionType, public llvm::FoldingSetNode {
+  enum class OverrideResult {
+    NoAction,
+    Warn,
+    Merge // Merge missing effect from base to derived
+  };
+
+  /// Return true if adding or removing the effect as part of a type conversion
+  /// should generate a diagnostic.
+  bool shouldDiagnoseConversion(QualType SrcType,
+                                const FunctionEffectsRef &SrcFX,
+                                QualType DstType,
+                                const FunctionEffectsRef &DstFX) const;
+
+  /// Return true if adding or removing the effect in a redeclaration should
+  /// generate a diagnostic.
+  bool shouldDiagnoseRedeclaration(const FunctionDecl &OldFunction,
+                                   const FunctionEffectsRef &OldFX,
+                                   const FunctionDecl &NewFunction,
+                                   const FunctionEffectsRef &NewFX) const;
+
+  /// Return true if adding or removing the effect in a C++ virtual method
+  /// override should generate a diagnostic.
+  OverrideResult shouldDiagnoseMethodOverride(
+      const CXXMethodDecl &OldMethod, const FunctionEffectsRef &OldFX,
+      const CXXMethodDecl &NewMethod, const FunctionEffectsRef &NewFX) const;
+};
+
+/// Support iteration in parallel through a pair of FunctionEffect and
+/// FunctionEffectCondition containers.
+template <typename Container> class FunctionEffectIterator {
+  const Container &Outer;
+  size_t Idx;
+
+public:
+  FunctionEffectIterator(const Container &O, size_t I) : Outer(O), Idx(I) {}
+  bool operator==(const FunctionEffectIterator &Other) const {
+    return Idx == Other.Idx;
+  }
+  bool operator!=(const FunctionEffectIterator &Other) const {
+    return Idx != Other.Idx;
+  }
+
+  // prefix increment
+  FunctionEffectIterator operator++() {
+    ++Idx;
+    return *this;
+  }
+
+  FunctionEffectWithCondition operator*() const {
+    const bool HasConds = !Outer.Conditions.empty();
+    return FunctionEffectWithCondition{Outer.Effects[Idx],
+                                       HasConds ? Outer.Conditions[Idx]
+                                                : FunctionEffectCondition()};
+  }
+};
+
+/// An immutable set of FunctionEffects and possibly conditions attached to
+/// them. The effects and conditions reside in memory not managed by this object
+/// (typically, trailing objects in FunctionProtoType, or borrowed references
+/// from a FunctionEffectSet).
+///
+/// Invariant: there is never more than one instance of any given effect.
+class FunctionEffectsRef {
+  ArrayRef<FunctionEffect> Effects;
+
+  // The array of conditions is either empty or has the same size
+  // as the array of effects.
+  ArrayRef<FunctionEffectCondition> Conditions;
+
+public:
+  /// Extract the effects from a Type if it is a function, block, or member
+  /// function pointer, or a reference or pointer to one.
+  static FunctionEffectsRef get(QualType QT);
+
+  FunctionEffectsRef() = default;
+
+  // The arrays are expected to have been sorted by the caller.
```

Comment:
Asserting is easy with `std::is_sorted()`.

It's a little sad though because this constructor is used whenever someone wants to look at `FunctionProtoType`'s effects. Another approach would be to make the constructor private, and add `FunctionProtoType` and `FunctionEffectSet` as friends, making them both responsible for the invariant.

Reply:
> Asserting is easy with `std::is_sorted()`.

Ah yeah; I candidly forgot it existed because I almost never have a reason to use it.

> add FunctionProtoType and FunctionEffectSet as friends, making them both responsible for the invariant.

If that doesn’t make the rest of the code much more complicated then I’d go with that; I’m just worried about a function (in this case a constructor) that has invariants that are supposed to be checked by the caller, but which anyone can just call without checking them.

Reply:
It was workable to make the constructor private and the friends responsible for the invariant.